### DISCUSSION TITLE
Add C# Extensions for Accessibility APIs

### DISCUSSION THREAD

**Discussion by TheCodeTraveler** [CommunityToolkit/Maui.Markup#58] [New Feature Discussions]
Let's add C# Extension Methods for the following Accessibility APIs:
- `Element`
  - `AutomationId`
- `SemanticProperties`
  - `SemanticProperties.DescriptionProperty`
  - `SemanticProperties.HeadingLevelProperty`
  - `SemanticProperties.HintProperty`
- `AutomationProperties`
  - `AutomationProperties.ExcludedWithChildrenProperty`
  - `AutomationProperties.HelpTextProperty`
  - `AutomationProperties.IsInAccessibleTreeProperty`
  - `AutomationProperties.LabeledByProperty`
  - `AutomationProperties.NameProperty`

**Comment by pictos** [comment]
I approve this feature ✅

**Reply by jfversluis** [reply]
Me too! This explains why I couldn't find this just now :D

**Reply by bijington** [reply]
I also approve

**Comment by mrlacey** [comment]
Apart from the following, it seems all of these have been implemented.

- AutomationProperties.HelpTextProperty
- AutomationProperties.LabeledByProperty
- AutomationProperties.NameProperty

Is there a reason not to include these?

**Reply by bijington** [reply]
These weren't added based on the documentation at https://learn.microsoft.com/en-gb/dotnet/maui/fundamentals/accessibility?view=net-maui-7.0#name stating that they have been superseded by `SemanticProperties`.

**Reply by bijington** [reply]
Perhaps we should mark this reason explicitly somewhere

**Reply by mrlacey** [reply]
I thought this was probably the case but there's no such superseded notice on the docs pages [for](https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.controls.automationproperties.helptextproperty?view=net-maui-7.0&source=docs) [those](https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.controls.automationproperties.labeledbyproperty?view=net-maui-7.0) [properties](https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.controls.automationproperties.nameproperty?view=net-maui-7.0). That's where I'd expect to see it recorded.


>  Perhaps we should mark this reason explicitly somewhere

I think that as it's now documented here it should be sufficient. If you wanted to be really thorough, we could add them with an `Obsolete` attribute to point people towards using the SematincProperties instead.

**Reply by bijington** [reply]
Paging @jfversluis do you know whether these `AutomationProperties` really are deprecated or whether the docs should mention them being superseded by `SemanticProperties`.

On the idea of adding them and marking as Obsolete I think I like the idea. It certainly makes it more discoverable and means we don't have to say RTFM. @brminnick do you have any thoughts?

**Reply by TheCodeTraveler** [reply]
> If you wanted to be really thorough, we could add them with an Obsolete attribute to point people towards using the SematincProperties instead 

On the surface, it does seem silly to add a new API with the `[Obsolete]` attribute, but I actually think this is a good idea and makes a lot of sense!

**Reply by jfversluis** [reply]
@rachelkang @pureween are the accessibility wizards!
